File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewGroupManager.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.uimanager;

import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.Nullable;
import com.facebook.react.bridge.UiThreadUtil;
import java.util.List;
import java.util.WeakHashMap;

/** Class providing children management API for view managers of classes extending ViewGroup. */
public abstract class ViewGroupManager<T extends ViewGroup>
    extends BaseViewManager<T, LayoutShadowNode> implements IViewManagerWithChildren {

  private static WeakHashMap<View, Integer> mZIndexHash = new WeakHashMap<>();

  @Override
  public LayoutShadowNode createShadowNodeInstance() {
    return new LayoutShadowNode();
  }

  @Override
  public Class<? extends LayoutShadowNode> getShadowNodeClass() {
    return LayoutShadowNode.class;
  }

  @Override
  public void updateExtraData(T root, Object extraData) {}

  public void addView(T parent, View child, int index) {
    parent.addView(child, index);
  }

  /**
   * Convenience method for batching a set of addView calls Note that this adds the views to the
   * beginning of the ViewGroup
   *
   * @param parent the parent ViewGroup
   * @param views the set of views to add
   */
  public void addViews(T parent, List<View> views) {
    UiThreadUtil.assertOnUiThread();

    for (int i = 0, size = views.size(); i < size; i++) {
      addView(parent, views.get(i), i);
    }
  }

  public static void setViewZIndex(View view, int zIndex) {
    mZIndexHash.put(view, zIndex);
  }

  public static @Nullable Integer getViewZIndex(View view) {
    return mZIndexHash.get(view);
  }

  public int getChildCount(T parent) {
    return parent.getChildCount();
  }

  public View getChildAt(T parent, int index) {
    return parent.getChildAt(index);
  }

  public void removeViewAt(T parent, int index) {
    UiThreadUtil.assertOnUiThread();

    parent.removeViewAt(index);
  }

  public void removeView(T parent, View view) {
    UiThreadUtil.assertOnUiThread();

    for (int i = 0; i < getChildCount(parent); i++) {
      if (getChildAt(parent, i) == view) {
        removeViewAt(parent, i);
        break;
      }
    }
  }

  public void removeAllViews(T parent) {
    UiThreadUtil.assertOnUiThread();

    for (int i = getChildCount(parent) - 1; i >= 0; i--) {
      removeViewAt(parent, i);
    }
  }

  /**
   * Returns whether this View type needs to handle laying out its own children instead of deferring
   * to the standard css-layout algorithm. Returns true for the layout to *not* be automatically
   * invoked. Instead onLayout will be invoked as normal and it is the View instance's
   * responsibility to properly call layout on its children. Returns false for the default behavior
   * of automatically laying out children without going through the ViewGroup's onLayout method. In
   * that case, onLayout for this View type must *not* call layout on its children.
   */
  @Override
  public boolean needsCustomLayoutForChildren() {
    return false;
  }

  /**
   * Returns whether or not this View type should promote its grandchildren as Views. This is an
   * optimization for Scrollable containers when using Nodes, where instead of having one ViewGroup
   * containing a large number of draw commands (and thus being more expensive in the case of an
   * invalidate or re-draw), we split them up into several draw commands.
   */
  public boolean shouldPromoteGrandchildren() {
    return false;
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewGroupManager.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.uimanager

import android.view.View
import android.view.ViewGroup
import androidx.core.view.doOnDetach
import com.facebook.common.logging.FLog
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.UiThreadUtil
import java.util.WeakHashMap

@Suppress("DEPRECATION")
public abstract class ViewGroupManager<T : ViewGroup>
@JvmOverloads
constructor(reactContext: ReactApplicationContext? = null) :
    BaseViewManager<T, LayoutShadowNode>(reactContext), IViewGroupManager<T> {

    public override fun createShadowNodeInstance(): LayoutShadowNode = LayoutShadowNode()

    public override fun getShadowNodeClass(): Class<out LayoutShadowNode> =
        LayoutShadowNode::class.java

    public override fun updateExtraData(root: T, extraData: Any): Unit = Unit

    // parent: childIndex[] - Used when we can't immediately add a view
    protected val operationsMap: WeakHashMap<T, MutableMap<Int, Boolean>> = WeakHashMap()

    public fun addViewSafely(parent: T, child: View, index: Int, callback: () -> Unit) {
        UiThreadUtil.assertOnUiThread()

        if (child.parent == null) {
            callback()
            return
        }

        operationsMap.getOrPut(parent) {
            mutableMapOf()
        }[index] = true

        // When the child-parent relation is removed, onDetachedFromWindow will be called.
        // Its important to wait for detaching as the view might be in a transition, and isn't removed immediately.
        child.doOnDetach {
            // Looking at how endViewTransition is implemented, dispatchDetachedFromWindow
            // gets called _before_ the parent relation is removed, so we need to post this to the end of the frame:
            child.post {
                if(operationsMap.remove(parent) == null) {
                    // The addView operation was already countered by a removeView operation while we were waiting
                    FLog.w("ReactClippingViewManager", "Tried to add a view to a parent after the child was detached, but a remove operation was already enqueued")
                    return@post
                }
                FLog.w("ReactClippingViewManager", "addView(): ${child::class.java.simpleName} had a parent, removed from previous parent and after onDetach adding to new parent $parent")
                callback()
            }
        }

        // With the detach listener in place, we can now remove the view from the previous parent:
        // Note: This call here is potentially redundant, as SurfaceMountingManager.kt is already removing it
        (child.parent as? ViewGroup)?.removeView(child)
    }

    public override fun addView(parent: T, child: View, index: Int): Unit =
        addViewSafely(parent, child, index) { parent.addView(child, index) }

    /**
     * Convenience method for batching a set of addView calls Note that this adds the views to the
     * beginning of the ViewGroup
     *
     * @param parent the parent ViewGroup
     * @param views the set of views to add
     */
    public fun addViews(parent: T, views: List<View>) {
        UiThreadUtil.assertOnUiThread()
        views.forEachIndexed { i, view -> addView(parent, view, i) }
    }

    public override fun getChildCount(parent: T): Int = parent.childCount

    public override fun getChildAt(parent: T, index: Int): View? = parent.getChildAt(index)

    public override fun removeViewAt(parent: T, index: Int) {
        UiThreadUtil.assertOnUiThread()
        parent.removeViewAt(index)
        operationsMap[parent]?.remove(index)
    }

    /**
     * Expo overrides this function GroupViewManagerWrapper.kt`, which is a replacement view manager
     * adding support for delegates receiving callbacks whenever one of the methods in the view
     * manager are called.
     */
    public open fun removeView(parent: T, view: View) {
        UiThreadUtil.assertOnUiThread()

        for (i in 0 until getChildCount(parent)) {
            if (getChildAt(parent, i) === view) {

                removeViewAt(parent, i)
                break
            }
        }
    }

    /**
     * Returns whether this View type needs to handle laying out its own children instead of deferring
     * to the standard css-layout algorithm. Returns true for the layout to *not* be automatically
     * invoked. Instead onLayout will be invoked as normal and it is the View instance's
     * responsibility to properly call layout on its children. Returns false for the default behavior
     * of automatically laying out children without going through the ViewGroup's onLayout method. In
     * that case, onLayout for this View type must *not* call layout on its children.
     */
    public override fun needsCustomLayoutForChildren(): Boolean = false

    public companion object {
        private val zIndexHash: WeakHashMap<View, Int> = WeakHashMap()

        @JvmStatic
        public fun setViewZIndex(view: View, zIndex: Int): Unit = zIndexHash.set(view, zIndex)

        @JvmStatic public fun getViewZIndex(view: View?): Int? = zIndexHash[view]
    }
}
